sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala

```diff
@@ -595,6 +598,22 @@ object BooleanSimplification extends Rule[LogicalPlan] with PredicateHelper {
 }
 
 /**
+ * Combines all adjacent [[Union]] and [[Unions]] operators into a single [[Unions]].
+ */
+object CombineUnions extends Rule[LogicalPlan] {
+  private def collectUnionChildren(plan: LogicalPlan): Seq[LogicalPlan] = plan match {
+    case Union(l, r) => collectUnionChildren(l) ++ collectUnionChildren(r)
```

Comment:
you should write this without using recursion to avoid stack overflow.

Reply:
Hi @marmbrus Could I ask you a question regarding your comment here? I don't understand 
the following sentence. Could you give me an example? Thanks!
> i.e. we'd miss cached data unioned with other data